pageserver/src/tenant/timeline.rs

```diff
@@ -631,6 +633,66 @@ impl Timeline {
             return Ok(());
         }
 
+        // retry two times to allow first round to find layers which need to be downloaded, then
+        // download them, then retry compaction
+        for round in 0..ROUNDS {
+            // should we error out with the most specific error?
+            let last_round = round == ROUNDS - 1;
+
+            let res = self.compact_without_ondemand_downloads(ctx).await;
+
+            let rls = match res {
+                Ok(()) => return Ok(()),
+                Err(CompactionError::Phase1(Phase1Error::DownloadRequired(rls))) if !last_round => {
+                    // this can be done at most one time before exiting, waiting
+                    rls
+                }
+                Err(CompactionError::Phase1(e @ Phase1Error::DownloadRequired(_))) => {
+                    // it is possible that this keeps happening, because we are battling
+                    // some other process which evicts the layers we download; we should
+                    // retry quite soon anyways.
+                    return Err(e.into());
+                }
+                Err(CompactionError::Phase1(Phase1Error::Other(e)) | CompactionError::Other(e)) => {
+                    return Err(e);
+                }
+            };
+
+            // this path can be visited in the second round of retrying, if first one found that we
+            // must first download some remote layers
+            let total = rls.len();
+
+            let mut downloads = rls
+                .into_iter()
+                .map(|rl| self.download_remote_layer(rl))
+                .collect::<futures::stream::FuturesUnordered<_>>();
+
+            let mut failed = 0;
+
+            while let Some(res) = downloads.next().await {
```

Comment:
- Add TODO that this `.await` should be sensitive to `ctx` cancellation once that is supported.
- In the meantime, make it sensitive to `task_mgr` cancellation (I think even compaction via mgmt API runs in a task_mgr task nowadays, should work fine)

Reply:
Good point. Implemented racing against cancellation in 297b1a70.

Do you want an extra todo as well? Note that it doesn't seem like there are any other cancellation checks in compaction...

Reply:
Extra todo assumed not necessary, resolving.